extensions-core/hdfs-storage/pom.xml

```diff
@@ -152,10 +152,6 @@
             </exclusion>
           </exclusions>
         </dependency>
-        <dependency>
```

Comment:
why is this removed, this will break the HDFS module.

Reply:
@jihoonson FYI one of the use case of HDFS module is to load data from S3A using HDFS APIs thus the need for drivers.

Reply:
Why does this break the HDFS module? AFAIK, S3A is in `hadoop-aws.jar` which needs to be added to the extension classpath. Druid's hdfs-storage doesn't use any API of `aws-java-sdk`.

Reply:
but i see this 
```
<exclusions>
            <exclusion>
              <groupId>com.amazonaws</groupId>
              <artifactId>aws-java-sdk</artifactId>
            </exclusion>
          </exclusions>
```
thus i think why it is needed...

Reply:
i think what you are suggestion will lead to https://hadoop.apache.org/docs/stable/hadoop-aws/tools/hadoop-aws/index.html#ClassNotFoundException:_com.amazonaws.services.s3.AmazonS3Client

Reply:
I think it would be better to add a proper version of aws-java-sdk to the extension classpath and this is easy if `pull-deps` utility is used. It downloads all necessary dependencies at once. Forcing a fixed version of aws-java-sdk for various hadoop versions might cause many potential version conflicts.

Reply:
@jihoonson pull-deps is not an option for security concerns, a lot of Druid cluster cannot access the internet for obvious security reasons. not sure what is the concern with having HDFS module packaged with the necessary dependencies.

Reply:
> not sure what is the concern with having HDFS module packaged with the necessary dependencies.

Having dependencies for hadoop-aws and aws-java-sdk here doesn't provide the necessary dependencies. It means that the hadoop-aws of `${hadoop.compile.version}` and aws-java-sdk of `${aws.sdk.bundle.version}` are included in the classpath. Obviously this can make potential conflicts with various hadoop versions. Even for secure hadoop clusters, they need to install hadoop-aws and aws-java-sdk in some ways even if that is not directly downloading from internet.

Reply:
Changed the scope of hadoop-aws to `provided` as well.

Reply:
@jihoonson not sure what you mean by 
 >   Even for secure hadoop clusters, they need to install hadoop-aws and aws-java-sdk in some ways even if that is not directly downloading from internet.

Nothing needs to be installed if all the dependency is part of HDFS module directory. This is needed for historical as well. Same thing for Hadoop common it is part of the module.

Reply:
@b-slim probably my explanation was not detailed enough. Sorry for that.

As you know, `druid-hdfs-storage` module is an extension for using HDFS as deep storage for druid. To do so, druid calls some HDFS APIs which requires to include some Hadoop dependencies like `hadoop-common` and `hadoop-client` by default. 

However, `hadoop-aws` is somewhat different because it is used in HDFS side and is not needed in druid side. If users want to make their HDFS working with s3, they should have already installed `hadoop-aws` in somewhere when HDFS starts up. Then, they can add `hadoop-aws` to Druid's classpath easily (this is need to push/pull segments to/from HDFS). If they don't use HDFS but want to use s3, they're encouraged to use `druid-s3-extensions` rather than `druid-hdfs-storage`. 

`aws-java-sdk` is used in `hadoop-aws`, but not in `druid-hdfs-storage`. Since the version of `aws-java-sdk` is different according to which version of `hadoop-aws` is used, it should be provided along with `hadoop-aws`. If we fix the version of `aws-java-sdk` here, it might mismatch with the one used in `hadoop-aws` and cause operation failures when trying to access to s3.

What do you think?

Reply:
@jihoonson i don't agree, i don't see why this should happen, i think druid-hdfs-storge should be able to package all its dependency and same for druid-s3-extensions, in the end only one extension is loaded thus i don't see how the conflict will happen unless you are telling me that Druid core class loader contains aws-java-sdk jars even if the user is not using druid-s3-extensions which is wrong IMO since Druid core does not need this to work. Does that make sense?

Reply:
" in the end only one extension is loaded" - is not entirely correct, the historicals may want to load both the extensions and independent segments can be loaded from different deep storage. e.g. A user migrated the deep storage from hdfs to s3 but choose to keep older segments in hdfs only. 

However,  if both extensions load their dependencies in their own class loaders, there should not be any conflict.

Reply:
@nishantmonu51 I don't think Druid allow you to load both at the same time since the is one property string that does the binding. Today if you load both you have a dependency injection issue. Unless I am missing something?

Reply:
Ok. First of all, I'm telling you that conflicts doesn't exist in druid, but exist when druid interacts with hadoop. In druid, there are two modules, `indexing-hadoop` and `hdfs-storage`. In both modules, druid calls some hdfs apis. Another thing which should be noted is that druid and hadoop have many dependencies on the same library but different versions. So, when druid calls some hdfs apis, the version of those libraries should be matched or at least compatible. If not, it causes conflicts.

> i think druid-hdfs-storge should be able to package all its dependency and same for druid-s3-extensions

@b-slim I agree on that we will be happy if we can do this, but I think, in this case, druid-hdfs-storage can't package all its dependencies because the version of those dependencies varies with different hadoop versions. If we can't achieve this, I think it would be better for users to put proper libraries because at least users can match the library versions by themselves. If the library version is fixed, there is no way to avoid conflicts.

> Druid core class loader contains aws-java-sdk jars even if the user is not using druid-s3-extensions

`aws-module` is one of Druid's core module and it has a dependency on `aws-java-sdk`. 

> FYI one of the use case of HDFS module is to load data from S3A using HDFS APIs thus the need for drivers.

Finally, I've tested the current master with hadoop 2.7.5 by using s3 as a deep storage via hdfs, and it doesn't work due to a removed API in aws-java-sdk. I've test both S3A and S3N. @b-slim is this use case one of your customers? Are they using open source druid and hadoop or some version of hortonworks? If it's the latter case, I think this patch should be fine.

Reply:
@jihoonson I have a question why Druid core need to load AWS stuff if it does not run or interact with AWS this seems to be a poor choice?

Reply:
That's a good question. I'm also not sure why it is, but it was there since 2015. Maybe we can also change it to be provided when loading an extension. I think it should be done in a follow-up pr if needed.